[DOCS] Adds description of analysis_stats object and its properties to GET DFA stats API docs

szabosteve opened the pull request:
This PR contains the documentation of the new `analysis_stats` object and its properties that are exposed in GET DFA stats.

Preview: http://elasticsearch_53881.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/get-dfanalytics-stats.html

szabosteve committed:
[DOCS] Adds description of analysis_stats object and its properties to GET DFA stats API docs.

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

elasticmachine commented:
Pinging @elastic/ml-core (:ml)

szabosteve committed:
[DOCS] Fixes typos.

szabosteve committed:
[DOCS] Adds hyperparameter related properties.

szabosteve committed:
[DOCS] Organizes content to collapsible sections.

szabosteve committed:
[DOCS] Adjusts indentation.

szabosteve commented:
@elasticmachine test this please

szabosteve committed:
[DOCS] Adds child attributes role.

szabosteve committed:
[DOCS] Adds role to analysis_stats.

szabosteve commented on docs/reference/ml/df-analytics/apis/get-dfanalytics-stats.asciidoc:
```diff
@@ -35,7 +35,7 @@ privileges:
 For more information, see <<security-privileges>> and <<built-in-roles>>.
 
-
+[role="child_attributes"]
```
```suggestion
```

szabosteve committed:
Update docs/reference/ml/df-analytics/apis/get-dfanalytics-stats.asciidoc

szabosteve committed:
[DOCS] Fixes markup.

dimitris-athanasiou commented:
There are a few hyperparameters being renamed. It shouldn't change their description. I'll notify here as soon as the renaming is done.

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+//Begin class_hyperparameters
+`hyperparameters`::::
+(object)
+An object containing the hyperparameters of the {classanalysis}.
```
```suggestion
An object containing the parameters of the {classanalysis}.
```
Also the section is called "hyperparameters" not all of these parameters are estimated in a hyperparameter  optimization procedure. Hence, I would use the word "parameters" here.

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+
+`eta`::::
+(double)
+The value of the eta hyperparameter.
```
Since `eta` can be specified explicitly in the request body, maybe you can cross-reference it here?

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+
+`max_trees`::::
+(integer)
+The maximum number of trees.
```
```suggestion
The maximum number of trees in the forest.
```

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+Determines the maximum number of splits for every feature that can occur in a 
+decision tree when the tree is trained.
+
+`regularization_depth_penalty_multiplier`::::
```
After https://github.com/elastic/ml-cpp/pull/1096 is through, it will be called `alpha`, it can be cross-referenced with analysis config.

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+Tree depth tolerance is used for calculating the tree depth penalty. This is a 
+soft limit, it can be exceeded.
+
+`regularization_tree_size_penalty_multiplier`::::
```
This parameter will be called `gamma`, the same as in the analysis config.

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+(double)
+Regularization factor to penalize deeper trees when training decision trees.
+
+`regularization_leaf_weight_penalty_multiplier`::::
```
This guy will be simply called `lambda`

valeriy42 commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,357 @@ tag::data-frame-analytics-stats[]
+`class_assignment_objective`::::
+(string)
+Defines whether class assignment maximizes the accuracy or the minimum recall 
+metric. Possible values are `accuracy` and `minimum_recall`.
```
```suggestion
metric. Possible values are `maximize_accuracy` and `maximize_minimum_recall`.
```
Schema will be updated.

valeriy42 reviewed (commented):
There will be changes in the schema https://github.com/elastic/ml-json-schemas-private/pull/11 to synchronize the `stats` section and the job configuration section. I added comments to the classification part. The changes in the regularization part are similar since the schema is mostly the same.

szabosteve committed:
Update docs/reference/ml/ml-shared.asciidoc

Co-Authored-By: Valeriy Khakhutskyy <1292899+valeriy42@users.noreply.github.com>

szabosteve committed:
Apply suggestions from code review

Co-Authored-By: Valeriy Khakhutskyy <1292899+valeriy42@users.noreply.github.com>

szabosteve commented:
@valeriy42 As per https://github.com/elastic/ml-json-schemas-private/pull/11, should I change `regularization_soft_tree_depth_limit` to `soft_tree_depth_limit` and `regularization_soft_tree_depth_tolerance` to `soft_tree_depth_tolerance`, too?
I addressed your suggestions except for the cross-references. I need to rework the PUT API request body section, too, I'll add the references here afterward.

szabosteve committed:
[DOCS] Addresses feedback.

valeriy42 commented:
> @valeriy42 As per [elastic/ml-json-schemas-private#11](https://github.com/elastic/ml-json-schemas-private/pull/11), should I change `regularization_soft_tree_depth_limit` to `soft_tree_depth_limit` and `regularization_soft_tree_depth_tolerance` to `soft_tree_depth_tolerance`, too?
Yes please, I missed this during my review.

valeriy42 reviewed (commented):
looks good, as you said @szabosteve two renamed properties are missing

szabosteve committed:
[DOCS] Modifies the name of two properties.

valeriy42 reviewed (approved):
LGTM

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+
+`eta_growth_rate_per_tree`::::
+(double)
+Specifies with what rate to increase `eta` for every new tree added to the 
```
Minor suggestion for clarity:
```suggestion
Specifies the rate at which the `eta` increases for each new tree that is added to the 
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+`eta_growth_rate_per_tree`::::
+(double)
+Specifies with what rate to increase `eta` for every new tree added to the 
+forest. The rate of 1.0 does not change `eta`, the rate of 1.05 increases `eta` 
```
I think you can likely simplify this as follows:
```suggestion
forest. For example, a rate of `1.05` increases `eta` 
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+`max_optimization_rounds_per_hyperparameter`::::
+(integer)
+A multiplier responsible for determining the maximum number of 
+hyperparameter optimization steps in the bayesian optimization procedure. 
```
I think this should be capitalized:
```suggestion
hyperparameter optimization steps in the Bayesian optimization procedure. 
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+(integer)
+A multiplier responsible for determining the maximum number of 
+hyperparameter optimization steps in the bayesian optimization procedure. 
+Maximum number of steps is determined as the number of undefined hyperparameters 
```
```suggestion
The maximum number of steps is determined based on the number of undefined hyperparameters 
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+
+`num_folds`::::
+(integer)
+Maximum number of folds for the cross-validation procedure.
```
Synching with definition above this one:
```suggestion
The maximum number of folds for the cross-validation procedure.
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+//Begin reg_hyperparameters
+`hyperparameters`::::
+(object)
+An object containing the paramters of the {reganalysis}.
```
```suggestion
An object containing the parameters of the {reganalysis}.
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+
+`eta_growth_rate_per_tree`::::
+(double)
+Specifies with what rate to increase `eta` for every new tree added to the 
```
Clarifying to match earlier similar definition:
```suggestion
Specifies the rate at which the `eta` increases for every new tree that is added to the 
```

lcawl commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,355 @@ tag::data-frame-analytics-stats[]
+
+`max_optimization_rounds_per_hyperparameter`::::
+(integer)
+A multiplier responsible for determining the maximum number of 
```
> A multiplier responsible ... 

This content seems to duplicate an earlier definition.  Can the content be re-used?

szabosteve replied:
@lcawl Yes, it can be. I single-sourced all the possible shared properties in a7bec16.

szabosteve committed:
Apply suggestions from code review

Co-Authored-By: Lisa Cawley <lcawley@elastic.co>

szabosteve committed:
[DOCS] Single-sources shared parameters.

dimitris-athanasiou commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,335 @@ tag::data-frame-analytics-stats[]
+tag::dfas-timestamp[]
+`timestamp`::::
+(date)
+The timestamp when the statistics returned in milliseconds since the epoch.
```
I wonder if we could substitue `returned` with a more suitable verb here. Perhaps `... when the statistics were reported`?

dimitris-athanasiou commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,335 @@ tag::data-frame-analytics-stats[]
+tag::dfas-iteration[]
+`iteration`::::
+(integer)
+The number of iterations on the {dfanalytics-job}.
```
I would rather say here "number of iterations in the classification analysis" instead of the DFA job, as iterations is not a concept that is relevant for all analyses.

dimitris-athanasiou commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,335 @@ tag::data-frame-analytics-stats[]
+tag::dfas-timing-stats-elapsed[]
+`elapsed_time`::::
+(integer)
+Runtime of the {dfanalytics-job} in milliseconds.
```
This should explain this is the runtime of the analysis itself, not the whole job's. Note the job involves other steps that take time (e.g. reindexing, etc.) that are not accounted for by this measurement.

dimitris-athanasiou commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,335 @@ tag::data-frame-analytics-stats[]
+tag::dfas-timing-stats-iteration[]
+`iteration_time`::::
+(integer)
+Runtime of the {dfanalytics-job} since the last iteration in milliseconds.
```
It might be simpler to explain this as "Runtime of the latest iteration of the classification analysis in milliseconds"

dimitris-athanasiou commented on docs/reference/ml/ml-shared.asciidoc:
```diff
@@ -507,6 +507,335 @@ tag::data-frame-analytics-stats[]
+tag::dfas-validation-loss-type[]
+`loss_type`::::
+(string)
+The name of the loss metric. For example, `binomial_logistic`.
```
`name` -> `type` ?

dimitris-athanasiou reviewed (commented):
Left a few recommendations. As I realized there is reuse in regression/classification, my suggestion to explicitly mention `classification analysis` does not stand. Perhaps simply `analysis` instead?

szabosteve committed:
[DOCS] Addresses feedback.

dimitris-athanasiou reviewed (approved):
LGTM

szabosteve commented:
**Backport commits:**
7.x: eeb23e9
7.7: 7f54f90
